src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/AbstractRealtimeTestResultAction.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2017 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.junitrealtimetestreporter;
+
+import hudson.AbortException;
+import hudson.Main;
+import hudson.model.Run;
+import hudson.tasks.test.AbstractTestResultAction;
+import hudson.tasks.test.TestResult;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.kohsuke.stapler.HttpRedirect;
+import org.kohsuke.stapler.StaplerProxy;
+
+abstract class AbstractRealtimeTestResultAction extends AbstractTestResultAction<AbstractRealtimeTestResultAction> implements StaplerProxy {
+
+    private static final Logger LOGGER = Logger.getLogger(AbstractRealtimeTestResultAction.class.getName());
+
+    protected TestResult result;
+    private transient long updated;
+
+    protected AbstractRealtimeTestResultAction() {}
+
+    protected abstract TestResult parse() throws IOException, InterruptedException;
+
+    @Override
+    public TestResult getResult() {
+        // Refresh every 1/100 of a job estimated duration but not more often than every 5 seconds
+        final long threshold = Math.max(5000, run.getEstimatedDuration() / 100);
```

Comment:
Some suggested improvements (but orthogonal to this PR):

* always run `parse` in case `result == null`
* run `parse` regardless of cache if `result.getTotalCount() == 0`
* refresh no less often than every 1m even if job is estimated to take >100m